### PR TITLE
Playlists v2: Refactors, touch ups + Queue Mode

### LINKED ISSUE #1477: Add a watch queue like on Youtube
**Is your feature request related to a problem? Please describe.**
When wanting to watch multiple videos after each other, one need to either add them to the "watch later" playlist or in another browser tab (on desktop).

**Describe the solution you'd like**
Instead, a watch queue like on YouTube could be added. It could look the same as when playing a playlist, but more integrated into normal viewing. Basically just copy what YouTube does, or something similar.

**Additional context**
Youtube-queue-odysee-request
Just like this

### LINKED ISSUE #1542: Improve Playlist user experience and make functionality more intuitive
Currently the UX is for Playlists is hard to follow and quite unintuitive in some places. Per our call there are a lot of changes to make to make the feature more user friendly and accessible

### PR DESCRIPTION
**Pull request by saltrafael**
## Fixes

Issue Number: closes #1542, closes #1477

https://salt.odysee.com

### CODE REVIEW

**Comment by infinite-persistence** [comment]
Not critical and only if you have time -- a quick summary of changes for 1542 to help guide the review would be nice.

File: flow-typed/Collections.js [hunk 1]
```diff
@@ -14,14 +14,15 @@ declare type CollectionState = {
   pending: CollectionGroup,
   edited: CollectionGroup,
```
Comment (infinite-persistence) [inline flow-typed/Collections.js:15]: Placeholders don't look right

https://user-images.githubusercontent.com/64950861/171437237-25f21f5c-613e-4146-bcf6-50308029c685.mp4

**Review by infinite-persistence — CHANGES_REQUESTED** [review]

File: ui/component/router/index.js [hunk 2]
```diff
@@ -1,17 +1,27 @@
...
+
+const PLAYLIST_PATH = getPathForPage(PAGES.PLAYLIST);
 
 const select = (state, props) => {
-  const { uri } = props;
+  let { uri } = props;
+  const { pathname } = state.router.location;
+
+  if (pathname.startsWith(PLAYLIST_PATH)) {
+    const collectionId = pathname.replace(PLAYLIST_PATH, '');
```
Comment (infinite-persistence) [inline ui/component/router/index.js:22]: --- Maybe ---
How are these playlist URLs shared?  Can't seem to find it in the share dialog .... the share dialog always includes the channel name.

Usually, a shared link will include a referrer ID, redirect param (like sunset, etc.).  I think this will not parse the `collectionId` correctly for these cases.

File: ui/redux/selectors/collections.js [hunk 3]
```diff
@@ -232,6 +239,11 @@ export const makeSelectNameForCollectionId = (id: string) =>
     return (collection && collection.name) || '';
   });
 
+export const makeSelectUpdatedAtForCollectionId = (id: string) =>
+  createSelector(makeSelectCollectionForId(id), (collection) => {
+    return (collection && collection.updatedAt) || '';
+  });
+
```
Comment (infinite-persistence) [inline ui/redux/selectors/collections.js:246]: No more `makeSelect*` pattern for new selectors -- it doesn't really cache -- it creates a new cache each time and never re-uses it. so it somewhat a memory leak.  If you do an infinite for-loop that calls a `makeSelect*`, you will see the memory usage goes GB until garbage collection kicks in during idle.

Do `select*(state, id)` instead, and can make it non-caching for simplicity, until caching is really needed.  Inside it, you can call existing `makeSelect` selectors if you do not wish to clean those up yet.

File: ui/component/router/index.js (hunk 2 above)
Comment (saltrafael) [reply ui/component/router/index.js:22]: This was intended for this page, ex: https://salt.odysee.com/$/playlist/73c2a30109229b2f262574cb7881f265f3284475 , vs the current: https://odysee.com/$/list/4df6cc75f73f220cc9b479d789b06702c0eede1e (page title is just "Odysee")

Comment (infinite-persistence) [reply ui/component/router/index.js:22]: Yup.  The concern is just whether the client, browser, or anything will add extra stuff after that. Some examples for other types of URLs where stuff is appended.
- https://odysee.com/?&sunset=lbrytv
- https://odysee.com/@SomeOrdinaryGamers:a?r=3vzdBMnXEhDiByz2x7R1RkCmUw1Hwk82

Comment (saltrafael) [reply ui/component/router/index.js:22]: Oh I see the replace is the problem now my bad

Comment (saltrafael) [reply ui/component/router/index.js:22]: just realized this wouldn't happen since it is using the value on `pathname` so it already strips other queries

File: ui/scss/component/section.scss [hunk 4]
```diff
@@ -125,6 +125,21 @@
   font-size: var(--font-body);
 }
 
+.section__subtitle--small {
+  @extend .card__subtitle;
+  margin: 0px;
+  font-size: var(--font-xsmall);
+}
```
Comment (infinite-persistence) [inline ui/scss/component/section.scss:132]: From BEM perspective, a modifier must always modify something, not used standalone.  When the rule is followed, there is no need to `@extend`. There are a few incorrect patterns like this, but let's clean up moving forward.

- Remove `@extend .card__subtitle;`
- At the client side, use `section__subtitle section__subtitle--small`

File: ui/redux/reducers/collections.js [hunk 5]
```diff
@@ -3,35 +3,37 @@ import { handleActions } from 'util/redux-utils';
 import * as ACTIONS from 'constants/action_types';
 import * as COLS from 'constants/collections';
 
-const getTimestamp = () => {
-  return Math.floor(Date.now() / 1000);
-};
-
```
Comment (infinite-persistence) [inline ui/redux/reducers/collections.js:9]: :exclamation: Changing to just `Date.now()` creates a discrepancy with LBRY Desktop.  The timestamp changes between seconds and milliseconds as I switch from Desktop to Odysee, since this is synced to the wallet.

Comment (infinite-persistence) [reply ui/redux/reducers/collections.js:9]: I think the wallet data uses seconds for `updatedAt`.  Would changing this to ms causes the sync merge to break?

File: ui/redux/reducers/collections.js [hunk 6]
```diff
@@ -170,13 +161,17 @@ const collectionsReducer = handleActions(
...
+
       return {
         ...state,
         edited: editedCollections || state.edited,
         unpublished: unpublishedCollections || state.unpublished,
-        builtin: builtinCollections || state.builtin,
-        saved: savedCollections || state.saved,
+        builtin: newBuiltin || state.unpublished,
       };
     },
```
Comment (infinite-persistence) [inline ui/redux/reducers/collections.js:176]: Some of my collections are not appearing in LBRY Desktop.

Is `newBuiltin` meant to erase the old version, and replace with new version that uses `Date.now()` as the timestamp?

- [ ] `builtin: newBuiltin || state.unpublished` -- unpublished is a typo?
  - In Redux, we return the original state reference when we know there is no change. In this case, `newBuiltIn` will always be a new object, so the `||` fallback will never be used.

Anyway, unclear what `newBuiltin` is for, but if it's to override old wallet data, I think we can't do that when LBRY Desktop is still in active use -- they will override each other.  If there is behavioral differences, it might be easier to create new lists and do a one-time action to port over if we detect the odysee version is null and lbry version is non-null.  Then, change the reducer to never touch the existing ones, and GUI will never display them.  

This is one idea to disconnect LBRY and Odysee lists properly, but I still think they should be in sync.
```
favorites
watchlater
favorites_odysee
watchlater_odysee
```

File: ui/component/claimPreview/view.jsx [hunk 7]
```diff
@@ -244,16 +252,22 @@ const ClaimPreview = forwardRef<any, {}>((props: Props, ref: any) => {
...
+      e.stopPropagation();
+      doUriInitiatePlay({ uri, collectionId, source: collectionId === 'queue' ? collectionId : undefined }, true, true);
+    } else {
+      if (onClick) {
+        onClick(e, claim, indexInContainer); // not sure indexInContainer is used for anything.
+      }
+      e.stopPropagation();
     }
-    e.stopPropagation();
   };
```
Comment (infinite-persistence) [inline ui/component/claimPreview/view.jsx:265]: --- Not a blocker, but consider in future ---
From a shared component design perspective, I think we should try to make the component as dumb as possible, only doing what's necessary to display the data. Specific actions like `doUriInitiatePlay` shouldn't be a `ClaimPreview`'s responsibility -- we can provide generic hooks and render functions, etc. to tie up these actions.

- Is it possible for the client (`playlistCard`) to define `ClaimList :: onClick` instead, and do the `e.preventDefault` and `doUriInitiatePlay` there?
  - I believe `onClick` is already propagated between `ClaimList` and `ClaimPreview`, so the drilling is already done.  And you should have all the collection params at the client side.  The only thing to figure how is if the tweaks in `navLinkProps` can have an alternative solution.
- We would then also not need `playItemsOnClick`

If `doUriInitiatePlay` is a very common action, then incorperating it would make sense, but probably not when it's for a very specific use-case.  It makes tweaking `handleNavLinkClick` harder in the future.

File: ui/component/fileRenderFloating/view.jsx [hunk 8]
```diff
@@ -374,97 +394,133 @@ export default function FileRenderFloating(props: Props) {
...
+                  </div>
+
+                  {playingCollection && collectionSidebarId !== collectionId && (
+                    <PlaylistCard
+                      id={collectionId}
+                      uri={uri}
+                      bodyOpen={playlistBodyOpen}
+                      enableCardBody={() => setPlaylistBody(!playlistBodyOpen)}
+                      customTitle={__('Now playing: --[Which Playlist is currently playing]--') + ' '}
+                      playItemsOnClick
```
Comment (infinite-persistence) [inline ui/component/fileRenderFloating/view.jsx:514]: `playItemsOnClick`

Do you anticipate a future `PlaylistCard` mounted somewhere else without `playItemsOnClick`?  If not, maybe we can remove this prop from `PlaylistCard`, and just enable `playItemsOnClick` on the ClaimList inside it?

File: flow-typed/Collections.js [hunk 9]
```diff
@@ -14,14 +14,21 @@ declare type CollectionState = {
...
   builtin: CollectionGroup,
```
Comment (infinite-persistence) [inline flow-typed/Collections.js:16]: `/$/playlists`
image

File: flow-typed/Collections.js [hunk 10]
```diff
@@ -14,14 +14,21 @@ declare type CollectionState = {
...
-  saved: Array<string>,
```
Comment (infinite-persistence) [inline flow-typed/Collections.js:17]: I'm still seeing less playlists than in production

https://odysee-workspace.slack.com/archives/C02G20Z2AEL/p1654661641103199?thread_ts=1654624401.430989&cid=C02G20Z2AEL

**Review by infinite-persistence — CHANGES_REQUESTED** [review]

File: flow-typed/Collections.js [hunk 11]
```diff
@@ -14,14 +14,21 @@ declare type CollectionState = {
...
   isResolvingCollectionById: { [string]: boolean },
```
Comment (infinite-persistence) [inline flow-typed/Collections.js:17]: --- non blocker ---
image

probably not meant to wrap

File: flow-typed/Collections.js [hunk 12]
```diff
@@ -14,14 +14,21 @@ declare type CollectionState = {
...
   error?: string | null,
```
Comment (infinite-persistence) [inline flow-typed/Collections.js:18]: Hovering shouldn't be the only way to add queue as it doesn't work in some platforms.
Non-blocker as I guess, can think of alternatives later if don't want to expand the content menu any further.

https://odysee-workspace.slack.com/archives/C02G20Z2AEL/p1654663481645619?thread_ts=1654624401.430989&cid=C02G20Z2AEL

File: ui/component/claimCollectionAddButton/view.jsx [hunk 13]
```diff
@@ -2,40 +2,33 @@
 import * as MODALS from 'constants/modal_types';
 import * as ICONS from 'constants/icons';
 import React from 'react';
-import Button from 'component/button';
-import classnames from 'classnames';
-import Tooltip from 'component/common/tooltip';
+import FileActionButton from 'component/common/file-action-button';
```
Comment (infinite-persistence) [inline ui/component/claimCollectionAddButton/view.jsx:5]: Nice consolidation

File: flow-typed/Collections.js (hunk 9 above)
Comment (infinite-persistence) [reply flow-typed/Collections.js:16]: Missing key while adding list

image

File: flow-typed/Collections.js [hunk 14]
```diff
@@ -14,14 +14,21 @@ declare type CollectionState = {
...
+  queue: Collection,
 };
```
Comment (infinite-persistence) [inline flow-typed/Collections.js:20]: - [ ] List cannot be opened in legacy apps?

This may or may not be an issue, but just wanted to be sure:

- If I create a list using this branch, I can see the list from production, but I cannot access it.  The url ends up being `https://odysee.com/?lid=adaadeff-88e1-485b-ab8c-62d5f037bfa1`

I understand there are URL changes ("list" to "playlist"), but from a data perspective, the app should just retrieve the ID and display it, right?  Trying to figure out why it would affect production or legacy apps -- how the app routes it shouldn't be impacted.

**Review by infinite-persistence — CHANGES_REQUESTED** [review]

File: ui/page/collection/internal/collectionActions/internal/playButton/index.js [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+        firstItem = claim.permanent_url;
+      }
+    });
+  }
+
+  return {
+    uri: firstItem,
+  };
+};
+
```
Comment (infinite-persistence) [inline ui/page/collection/internal/collectionActions/internal/playButton/index.js:27]: - [ ] Performance concerns + wrong function used

Performance should drop significantly here if you have a super-huge, since the state-to-map is a hot area and we are looping the entire array.

Looking at the intent, it shouldn't be using `map` for various wrong reasons.  Just use a for-loop so we can short-circuit when the entry is found.  If many undefined claims are expected at the top, then cache this inside a `createCachedSelector` (although I doubt this is the case).

<sub>https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map</sub>

image

File: ui/page/collection/internal/collectionActions/internal/playButton/index.js [hunk 16]
```diff
@@ -0,0 +1,28 @@
+import { connect } from 'react-redux';
+import { selectClaimForUri } from 'redux/selectors/claims';
+import { selectUrlsForCollectionId } from 'redux/selectors/collections';
+import ShuffleButton from './view';
```
Comment (infinite-persistence) [inline ui/page/collection/internal/collectionActions/internal/playButton/index.js:4]: There are a bunch of `ShuffleButton` from copy-pasting.  Rename these for clarity.

File: ui/component/claimSupportButton/view.jsx [hunk 17]
```diff
@@ -2,9 +2,7 @@
 import * as MODALS from 'constants/modal_types';
```
Comment (infinite-persistence) [inline ui/component/claimSupportButton/view.jsx:2]: - [ ] "Support" modal on own list incorrectly stuck at Fiat

Go to your own list and click "Support".  In production, it correctly shows "Boost" and LBC is selected, but in this branch, "Boost" is shown but only fiat is selectable.

image

File: ui/constants/collections.js [hunk 18]
```diff
@@ -1,5 +1,8 @@
+import * as ICONS from 'constants/icons';
+
 // ui
 export const ICON_SIZE = 12;
+export const PLACEHOLDER = __('My Awesome Playlist');
```
Comment (infinite-persistence) [inline ui/constants/collections.js:5]: Minor, but this will only be evaluated once, and will not update when Language is changed.
- The constant should be a regular string, and the `__()` macro moved to the calling side.

File: ui/constants/pageTitles.js [hunk 19]
```diff
@@ -22,7 +22,7 @@ export const PAGE_TITLE = {
   [PAGES.HELP]: 'Help',
   [PAGES.INVITE]: 'Invite',
   [PAGES.LIBRARY]: 'Purchases',
-  [PAGES.LISTS]: 'Lists',
+  [PAGES.PLAYLISTS]: 'Playlists',
```
Comment (infinite-persistence) [inline ui/constants/pageTitles.js:25]: In the router, I like it that `/$/lists` is still supported for backward compatibility, for users still using old links in their blogposts or whatever.

This side should follow suite, covering both.

File: ui/redux/reducers/collections.js (hunk 5 above)
Comment (saltrafael) [reply ui/redux/reducers/collections.js:9]: for some reason when I use the previous function here `/ 1000` it returns 1970 as date

Comment (infinite-persistence) [reply ui/redux/reducers/collections.js:9]: Is it something local to the machine? https://stackoverflow.com/questions/26642990/javascript-s-date-now-returning-0-zero.

File: ui/redux/actions/collections.js [hunk 20]
```diff
@@ -5,36 +5,34 @@ import Lbry from 'lbry';
...
       entry: {
-        id: uuid(), // start with a uuid, this becomes a claimId after publish
-        name: name,
-        updatedAt: getTimestamp(),
-        items: collectionItems || [],
-        sourceId: sourceId,
-        type: type,
+        id: id, // start with a uuid, this becomes a claimId after publish
+        createdAt: Date.now() / 1000,
+        updatedAt: Date.now() / 1000,
```
Comment (infinite-persistence) [inline ui/redux/actions/collections.js:33]: - [ ] We need to retain the original code of including `Math.floor`, because the wallet data for `updatedAt` is an integer.
- [ ] While calling `Date.now()` is quick succession would result in the same value most of the time, I don't think this is guaranteed.  Many factors like browser implementation and execution delays could make the values drift.  Usually, just stash the value in a variable before the dispatch object is called.

File: ui/redux/actions/collections.js [hunk 21]
```diff
@@ -372,26 +378,38 @@ export const doCollectionEdit = (collectionId: string, params: CollectionEditPar
...
+
+  return dispatch({
+    type: ACTIONS.COLLECTION_EDIT,
+    data: {
+      collectionKey,
+      collection: {
+        items: newItems,
+        id: collectionId,
+        name: params.name || collection.name,
+        updatedAt: Date.now() / 1000,
```
Comment (infinite-persistence) [inline ui/redux/actions/collections.js:401]: `Math.floor`

File: ui/util/time.js [hunk 22]
```diff
@@ -71,3 +71,5 @@ export function getTimeAgoStr(
 
   return __(str, { duration });
 }
+
+export const getCurrentTimeInMs = () => Math.floor(Date.now() / 1000);
```
Comment (infinite-persistence) [inline ui/util/time.js:75]: I think the function is incorrectly named.
`Date.now()` = milliseconds
Divide by 1000 = seconds

File: ui/redux/actions/collections.js [hunk 23]
```diff
@@ -333,18 +334,23 @@ export const doFetchItemsInCollection = (options: { collectionId: string, pageSi
   return doFetchItemsInCollections(newOptions, cb);
 };
 
-export const doCollectionEdit = (collectionId: string, params: CollectionEditParams) => async (
+export const doCollectionEdit = (collectionId: string, params: CollectionEditParams) => (
```
Comment (infinite-persistence) [inline ui/redux/actions/collections.js:337]: - [ ] Anonymous list becomes non-anonymous if updated

I made sure to make the channel-selector in the page to be anonymous, but the API call still included `channel_id`.  I think it is active the default channel ID (same as the avatar)

File: ui/component/autoplayCountdown/view.jsx [hunk 24]
```diff
@@ -155,7 +160,11 @@ function AutoplayCountdown(props: Props) {
               iconRight={skipPaid ? ICONS.WALLET : ICONS.REPLAY}
               onClick={() => {
                 setTimerCanceled(true);
-                doReplay();
+                if (skipPaid) {
+                  doOpenModal(MODALS.AFFIRM_PURCHASE, { uri, cancelCb: () => setTimerCanceled(false) });
```
Comment (infinite-persistence) [inline ui/component/autoplayCountdown/view.jsx:164]: - [ ] "Play previous" not vertically centered
- [ ] Wallet icon would be cleaner if on the left as well.

image

File: ui/component/autoplayCountdown/view.jsx [hunk 25]
```diff
@@ -155,7 +160,11 @@ function AutoplayCountdown(props: Props) {
...
+                } else {
+                  doReplay();
+                }
```
Comment (infinite-persistence) [inline ui/component/autoplayCountdown/view.jsx:167]: Cancelling the autoplay countdown of a paid item causing the playlist play button to be stuck

- Open http://localhost:9090/@eevblog:7/eevblog-1396-blackmagic-atem-mini:d?lid=acd1afb0e8a16ba534d523a3ec49f0b28ae95bc0 in incognito
- Click "Next", then cancel the autoplay countdown.
- Click on the 3rd video.  It is stuck in the Big Red Button mode.

File: flow-typed/content.js [hunk 26]
```diff
@@ -2,15 +2,16 @@
...
   history: Array<WatchHistory>,
   recommendationId: { [string]: string }, // claimId: recommendationId
   recommendationParentId: { [string]: string }, // claimId: referrerId
   recommendationUrls: { [string]: Array<string> }, // claimId: [lbryUrls...]
   recommendationClicks: { [string]: Array<number> }, // "claimId": [clicked indices...]
-  loopList?: { collectionId: string, loop: boolean },
-  shuffleList?: { collectionId: string, newUrls: Array<string> | boolean },
   // TODO: it's confusing for newUrls to be a boolean --------- ^^^
   // It can/should be '?Array<string>` instead -- set it to null, then clients
   // can cast it to a boolean. That, or rename the variable to `shuffle` if you
```
Comment (infinite-persistence) [inline flow-typed/content.js:17]: This comment can be removed since the code is gone, plus the new implementation down below is using `?{ }` :+1:

File: ui/component/autoplayCountdown/view.jsx [hunk 27]
```diff
@@ -155,7 +160,11 @@ function AutoplayCountdown(props: Props) {
               iconRight={skipPaid ? ICONS.WALLET : ICONS.REPLAY}
               onClick={() => {
                 setTimerCanceled(true);
-                doReplay();
+                if (skipPaid) {
```
Comment (infinite-persistence) [inline ui/component/autoplayCountdown/view.jsx:163]: - [ ] Autoplay next is broken -- it just replays itself

Steps:
- http://localhost:9090/@BrittanyVenti:0/they-banned-me-for-being-too-funky-fresh:a in incognito.  Autoplay-next should be on by default
- Drag towards the end and let it continue to the next video.  When the countdown is over, it replays itself rather than the next video.

File: ui/page/collection/internal/collectionActions/internal/playButton/view.jsx [hunk 28]
```diff
@@ -0,0 +1,29 @@
+// @flow
```
Comment (infinite-persistence) [inline ui/page/collection/internal/collectionActions/internal/playButton/view.jsx:1]: Missing padding

image

File: ui/component/claimMenuList/view.jsx [hunk 29]
```diff
@@ -172,12 +175,18 @@ function ClaimMenuList(props: Props) {
     // $FlowFixMe
     (contentClaim.value.stream_type === 'audio' || contentClaim.value.stream_type === 'video');
 
-  function handleAdd(source, name, collectionId) {
+  function handleAdd(claimIsInPlaylist, name, collectionId) {
     doToast({
-      message: source ? __('Item removed from %name%', { name }) : __('Item added to %name%', { name }),
+      message: claimIsInPlaylist ? __('Item removed from %name%', { name }) : __('Item added to %name%', { name }),
+      linkText: __('View All'),
+      linkTarget: `/list/${collectionId}`,
```
[the remark this answers is not in the document]
Comment (saltrafael) [reply ui/component/claimMenuList/view.jsx:182]: This has been changed to a `Start Playing` button instead of `View All`, let me know what do you think

[the remark this answers is not in the document]
Comment (infinite-persistence) [reply ui/component/claimMenuList/view.jsx:182]: Sounds good.

File: ui/component/autoplayCountdown/view.jsx (hunk 27 above)
Comment (saltrafael) [reply ui/component/autoplayCountdown/view.jsx:163]: Fixed ?

Comment (infinite-persistence) [reply ui/component/autoplayCountdown/view.jsx:163]: Just tried on salt.  Still happening.

Comment (saltrafael) [reply ui/component/autoplayCountdown/view.jsx:163]: And now?

File: ui/redux/actions/content.js [hunk 30]
```diff
@@ -151,47 +209,187 @@ export function doDownloadUri(uri: string) {
...
+        dispatch(
+          doUriInitiatePlay(
+            { uri: createNew ? uri : firstItemUri, collection: { collectionId } },
+            true,
+            true,
+            !floatingPlayerEnabled && push ? (url) => push(url) : undefined
+          )
+        );
+      }
+    };
```
Comment (infinite-persistence) [inline ui/redux/actions/content.js:345]: - [ ] :exclamation: I added a video into a private collection, and clicked "Start playing" from the toast. It ended up playing Watch Later, which wasn't the target.
- [ ] :exclamation: Create a new private playlist, then add 1 item into from the homepage. Start Playing doesn't work until there are 2 items in the list.

**Review by infinite-persistence — CHANGES_REQUESTED** [review]

File: ui/page/collection/internal/collectionActions/view.jsx [hunk 31]
```diff
@@ -0,0 +1,110 @@
...
+        {!collectionEmpty && <ShuffleButton collectionId={collectionId} />}
+
+        {!isBuiltin && (
+          <>
+            {uri && (
+              <>
+                {ENABLE_FILE_REACTIONS && <FileReactions uri={uri} />}
+                <ClaimSupportButton uri={uri} fileAction />
+                <ClaimRepostButton uri={uri} />
+                <ClaimShareButton uri={uri} fileAction webShareable />
```
Comment (infinite-persistence) [inline ui/page/collection/internal/collectionActions/view.jsx:60]: - [ ] :exclamation: `ClaimShareButton` is missing the `collectionId` parameter here (or maybe somewhere else), so the Social Share modal doesn't show the "Include playlist ID" widget.

image

File: ui/component/walletSendTip/view.jsx [hunk 32]
```diff
@@ -246,16 +246,16 @@ export default function WalletSendTip(props: Props) {
...
-  }, [activeTab, hasSelected, hasSelectedTab, setActiveTab]);
+  }, [activeTab, claimIsMine, hasSelected, hasSelectedTab, setActiveTab]);
 
   React.useEffect(() => {
     if (!hasSelectedTab && activeTab !== hasSelectedTab) {
-      setPersistentTab(activeTab);
+      setPersistentTab(claimIsMine ? TAB_BOOST : activeTab);
     }
-  }, [activeTab, hasSelectedTab, setPersistentTab]);
+  }, [activeTab, claimIsMine, hasSelectedTab, setPersistentTab]);
```
Comment (infinite-persistence) [inline ui/component/walletSendTip/view.jsx:258]: - [ ] Something's incomplete ... maybe the button label needs update too?

image

File: ui/redux/actions/collections.js [hunk 33]
```diff
@@ -333,20 +335,53 @@ export const doFetchItemsInCollection = (options: { collectionId: string, pageSi
...
+      let url = selectPermanentUrlForUri(state, uri);
+
+      if (!url) {
+        const claim = await dispatch(doGetClaimFromUriResolve(url));
+        if (claim) url = claim.permanent_url;
+      }
+
+      // $FlowFixMe
+      return uris.push(url);
+    });
```
Comment (infinite-persistence) [inline ui/redux/actions/collections.js:381]: - [ ] `await` in a `forEach` is undefined behavior. I have had issues recently trying to do something similar.
    - https://www.becomebetterprogrammer.com/javascript-foreach-async-await/#:~:text=Programmers%20use%20the%20forEach%20method,async%20and%20await%20with%20it.
- [ ] Also, what happens when the list is huge and every `url` happens to be undefined -- does it resolve and await one by one?

File: ui/redux/actions/content.js (hunk 30 above)
Comment (saltrafael) [reply ui/redux/actions/content.js:369]: still happens?

File: ui/component/walletSendTip/view.jsx (hunk 32 above)
Comment (saltrafael) [reply ui/component/walletSendTip/view.jsx:258]: still happens? I don't see it

File: ui/redux/actions/content.js (hunk 30 above)
Comment (infinite-persistence) [reply ui/redux/actions/content.js]: The second item is still happening

File: ui/component/walletSendTip/view.jsx (hunk 32 above)
Comment (infinite-persistence) [reply ui/component/walletSendTip/view.jsx:258]: Yes.  Doesn't happen in production, so feels like something related to this change.
- Uploads > Click "Support" from the context menu of a Post

File: ui/redux/actions/claims.js [hunk 34]
```diff
@@ -186,6 +186,21 @@ export function doResolveUri(
...
+      .then((response) =>
+        Object.values(response).forEach((resposne) => {
+          claim = resposne;
+        })
+      )
+      .catch((e) => {});
+
+    return claim;
+  };
+}
```
Comment (infinite-persistence) [inline ui/redux/actions/claims.js:202]: Few minor concerns from a (re)usability standpoint:
- [ ] The function assumes the caller has checked that we have not resolved before, because `response` would be `undefined` if previously resolved.  Can either fix by setting the parameter to `false`, or just document this gotcha behavior.
- [ ] `resposne` typo
- [ ] But why `forEach`?  Assuming that it is possible for `resolve` to return multiple entries for 1 `uri`, setting `claim` directly to the last entry would be a faster equivalent than going through everything.

File: ui/modal/modalCollectionCreate/internal/collectionCreate/index.js [hunk 35]
```diff
@@ -0,0 +1,19 @@
...
+import { selectPermanentUrlForUri } from 'redux/selectors/claims';
+import { doToast } from 'redux/actions/notifications';
+
+const select = (state, props) => {
+  const { uri } = props;
+
+  return {
+    // lists can only use permanent_url
+    uri: selectPermanentUrlForUri(state, uri),
+  };
```
Comment (infinite-persistence) [inline ui/modal/modalCollectionCreate/internal/collectionCreate/index.js:12]: - [ ] Is `uri` needed when creating private collection? It is also passed in from the calling side, so should clean up there as well if not needed.

File: ui/redux/actions/claims.js (hunk 34 above)
Comment (infinite-persistence) [reply ui/redux/actions/claims.js:202]: - [ ] I believe the first checkbox above is causing https://sentry.lbry.tech/organizations/odysee/issues/28402, where `doGetClaimFromUriResolve` is still returning `undefined` despite the effort to resolve.
 
- [ ] On a related note, https://sentry.lbry.tech/organizations/odysee/issues/28455 is also failing from a `null` claim and would probably require the same solution.  However, since it is `null`, does that mean the claim is already abandoned?
